core/src/main/java/org/apache/calcite/rel/rules/materialize/MaterializedViewRule.java

```diff
@@ -1426,6 +1448,61 @@ protected enum MatchModality {
     QUERY_PARTIAL
   }
 
+  /**
+   * Check if the materialization is valid for this rule.
+   * If it does, cache the materialization's RelTableRef and RelOptPredicateList.
+   */
+  protected class MaterializationMetadata {
+    private final Set<RelOptMaterialization> notValidMaterializations = new HashSet<>();
+    private final Map<RelOptMaterialization, Set<RelTableRef>> viewTableRefsMap = new HashMap<>();
+    private final Map<RelOptMaterialization, RelOptPredicateList>
+        viewPredicateListMap = new HashMap<>();
+
+    protected boolean isValidMaterialization(
+        RelMetadataQuery mq, RelOptMaterialization materialization) {
+      if (notValidMaterializations.contains(materialization)) {
+        return false;
+      } else if (viewTableRefsMap.containsKey(materialization)) {
+        return true;
```

Comment:
Should we use `viewTableRefsMap.containsKey(materialization) && viewPredicateListMap.containsKey(materialization)` to be more accurate?

Reply:
> Should we use `viewTableRefsMap.containsKey(materialization) && viewPredicateListMap.containsKey(materialization)` to be more accurate?

Indeed we should.Thanks for your suggestion and review!